rotkehlchen/chain/evm/decoding/rainbow/decoder.py

```diff
@@ -0,0 +1,417 @@
+            wrap_event: EvmTxReceiptLog,
+    ) -> None:
+        """
+        Decodes a swap transaction from native token to another token.
+        Check if the first log is deposit and the last is transfer log
+        """
+        # The fee amount is the difference of msg.value and the wrapped token amount
+        try:
+            swap_amount = asset_raw_value(event.amount, self.evm_inquirer.native_token)
+        except UnsupportedAsset:
+            log.error(
+                f'Failed to decode swap from native token to token'
+                f' for transaction {transaction.tx_hash.hex()} '
+                f'due to unsupported asset {event.asset.identifier}.',
+            )
+            return
+
+        fee_raw = swap_amount - int.from_bytes(wrap_event.data)
+
+        # Create fee event
+        fee_event = self._create_fee_event(
+            tx_hash=transaction.tx_hash,
+            timestamp=transaction.timestamp,
+            fee_raw=fee_raw,
+            fee_asset=self.evm_inquirer.native_token,
+            sender=sender,
+        )
+        decoded_events.append(fee_event)
+
+        return
+
+    def _maybe_decode_fee_from_swap_token_to_native(
+            self,
+            transaction: EvmTransaction,
+            decoded_events: list['EvmEvent'],
+            in_event: Optional['EvmEvent'],
+            out_event: Optional['EvmEvent'],
+            sender: ChecksumEvmAddress,
+            unwrap_event: EvmTxReceiptLog,
+    ) -> None:
+        """
+        Decodes the fee from a swap transaction involving token to native token.
+        Via the first transfer log transfer and the last withdrawn log
+        """
+        native_token = self.evm_inquirer.native_token
+
+        if in_event is None or out_event is None:
+            log.error(
+                f'Failed to decode fee from swap token to native token '
+                f'for transaction {transaction.tx_hash.hex()}.',
+            )
+            return
+
+        try:
+            in_amount = asset_raw_value(in_event.amount, native_token)
+        except UnsupportedAsset:
+            log.error(
+                f'Failed to decode fee from swap token to native token '
+                f'for transaction {transaction.tx_hash.hex()} '
+                f'due to unsupported asset {in_event.asset.identifier}.',
+            )
+            in_amount = 0
+
+        maybe_reshuffle_events(
+            ordered_events=[out_event, in_event],
+            events_list=decoded_events,
+        )
+
+        # The fee will be the withdraw data subtract the in_amount
+        fee_raw = int.from_bytes(unwrap_event.data) - in_amount
+
+        # Create the event for the fee
+        fee_event = self._create_fee_event(
+            tx_hash=transaction.tx_hash,
+            timestamp=transaction.timestamp,
+            fee_raw=fee_raw,
+            fee_asset=native_token,
+            sender=sender,
+        )
+        decoded_events.append(fee_event)
```

Comment:
have the fee also properly sorted

```suggestion
        # The fee will be the withdraw data subtract the in_amount
        fee_raw = int.from_bytes(unwrap_event.data) - in_amount

        # Create the event for the fee
        fee_event = self._create_fee_event(
            tx_hash=transaction.tx_hash,
            timestamp=transaction.timestamp,
            fee_raw=fee_raw,
            fee_asset=native_token,
            sender=sender,
        )
        decoded_events.append(fee_event)
        maybe_reshuffle_events(
            ordered_events=[out_event, in_event, fee_event],
            events_list=decoded_events,
        )
```